openedx/features/learner_profile/tests/views/test_learner_profile.py

```diff
@@ -138,27 +139,22 @@ def test_certificate_visibility(self, cert_mode):
         cert = self._create_certificate(cert_mode)
         cert.save()
 
-        request = RequestFactory().get('/url')
-        request.user = self.user
-        context = learner_profile_context(request, self.user.username, self.user.is_staff)
+        response = self.client.get('/u/{username}'.format(username=self.user.username))
 
-        self.assertTrue('card certificate-card mode-' + cert_mode in str(context['achievements_fragment'].content))
```

Comment:
Could you explain what this was doing and how it was passing without really working?  I'm curious for the future.  Do all tests have to have the full request/response in order to be useful?

Reply:
There were a lot of things going on here:
 - there was no ``__init__.py`` so Nose wouldn't find the test
 - the Waffle flag was not on
 - the test didn't render the page but instead was verifying the Mako context that would be passed to it. This wasn't bad because the fragment was rendered, but I thought it better to actually request the whole page.

I also don't like tests that use ``assertTrue`` because the only result you get back is that the result wasn't true. Using ``assertContains`` gives back a result that shows what was expected to be found and what the actual string contained.